Maple2.Server.Game/Manager/ShopManager.cs

```diff
@@ -0,0 +1,609 @@
+            foreach (ShopItem shopItem in itemList) {
+                Item? item = session.Item.CreateItem(shopItem.ItemId, shopItem.Rarity, shopItem.Quantity);
+                if (item == null) {
+                    continue;
+                }
+                shopItem.Item = item;
+                items.Add(shopItem.Id, shopItem.Clone());
+            }
+        } else {
+            foreach ((int id, ShopItem shopItem) in shop.Items) {
+                Item? item = session.Item.CreateItem(shopItem.ItemId, shopItem.Rarity, shopItem.Quantity);
+                if (item == null) {
+                    continue;
+                }
+                shopItem.Item = item;
+                items.Add(id, shopItem.Clone());
+            }
+        }
+
+        return items;
+    }
+
+    private CharacterShopData CreateShopData(Shop shop) {
+        if (!TryGetShopData(shop.Id, out CharacterShopData? data)) {
+            data = new CharacterShopData {
+                Interval = shop.RestockData!.Interval,
+                RestockTime = GetRestockTime(shop.RestockData!.Interval),
+                ShopId = shop.Id,
+            };
+
+            if (shop.RestockData.PersistantInventory) {
+                accountShopData[shop.Id] = data;
+            } else {
+                characterShopData[shop.Id] = data;
+            }
+        }
+
+        shop.RestockTime = data.RestockTime;
+
+        // We do not save shops with minute intervals to database
+        if (shop.RestockData!.Interval != ShopRestockInterval.Minute) {
+            using GameStorage.Request db = session.GameStorage.Context();
+            long ownerId = shop.RestockData.PersistantInventory ? session.AccountId : session.CharacterId;
+            data = db.CreateCharacterShopData(ownerId, data);
+            if (data == null) {
+                throw new InvalidOperationException($"Failed to create shop data {shop.Id} for character {session.CharacterId}");
+            }
+        }
+
+        return data;
+    }
+
+    private IDictionary<int, CharacterShopItemData> CreateShopItemData(Shop shop) {
+        Dictionary<int, CharacterShopItemData> itemData = new Dictionary<int, CharacterShopItemData>();
+        using GameStorage.Request db = session.GameStorage.Context();
+        long ownerId = shop.RestockData!.PersistantInventory ? session.AccountId : session.CharacterId;
+        foreach ((int id, ShopItem item) in shop.Items) {
+            var data = new CharacterShopItemData {
+                ShopId = shop.Id,
+                ShopItemId = id,
+                Item = item.Item,
+            };
+            if (shop.RestockData.Interval != ShopRestockInterval.Minute) {
+                data = db.CreateCharacterShopItemData(ownerId, data);
+                if (data == null) {
+                    continue;
+                }
+            }
+            itemData[id] = data;
+        }
+
+        if (shop.RestockData!.PersistantInventory) {
```

Comment:
`shop.RestockData?.PersistantInventory == true` seems safer, you should use `!` more as last resort